File: backend/app/schemas/payment.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class PaymentCreate(BaseModel):
    booking_id: uuid.UUID
    party_type: Literal["customer", "broker"]
    payment_date: date
    amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    tds_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    payment_method: str | None = None
    reference_number: str | None = None
    remarks: str | None = None


class PaymentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    booking_id: uuid.UUID
    customer_id: uuid.UUID | None
    broker_id: uuid.UUID | None
    party_type: str
    payment_date: date
    amount: Decimal
    tds_amount: Decimal
    payment_method: str | None
    reference_number: str | None
    remarks: str | None
    created_at: datetime


class LedgerEntry(BaseModel):
    entry_date: date
    booking_id: uuid.UUID | None = None
    payment_id: uuid.UUID | None = None
    booking_number: str | None = None
    description: str
    debit: Decimal
    credit: Decimal
    running_balance: Decimal


class LedgerResponse(BaseModel):
    party_id: uuid.UUID
    party_name: str
    party_type: Literal["customer", "broker"]
    outstanding_amount: Decimal
    entries: list[LedgerEntry]

File: frontend/src/services/payment.ts
import api from "./api";

export type PaymentParty = "customer" | "broker";

export interface Payment {
  id: string;
  booking_id: string;
  customer_id?: string;
  broker_id?: string;
  party_type: PaymentParty;
  payment_date: string;
  amount: number;
  tds_amount: number;
  payment_method?: string;
  reference_number?: string;
  remarks?: string;
  created_at: string;
}

export interface CreatePaymentPayload {
  booking_id: string;
  party_type: PaymentParty;
  payment_date: string;
  amount: number;
  tds_amount?: number;
  payment_method?: string;
  reference_number?: string;
  remarks?: string;
}

export interface LedgerEntry {
  entry_date: string;
  booking_id?: string;
  payment_id?: string;
  booking_number?: string;
  description: string;
  debit: number;
  credit: number;
  running_balance: number;
}

export interface Ledger {
  party_id: string;
  party_name: string;
  party_type: PaymentParty;
  outstanding_amount: number;
  entries: LedgerEntry[];
}

export const getPayments = async () => (await api.get<Payment[]>("/payments/")).data;
export const createPayment = async (data: CreatePaymentPayload) => (await api.post<Payment>("/payments/", data)).data;
export const deletePayment = async (id: string) => (await api.delete(`/payments/${id}`)).data;
export const getCustomerLedger = async (id: string) => (await api.get<Ledger>(`/ledgers/customers/${id}`)).data;
export const getBrokerLedger = async (id: string) => (await api.get<Ledger>(`/ledgers/brokers/${id}`)).data;
